servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TaskBasedAsyncCompletableOperator.java

```diff
@@ -292,27 +314,42 @@ public void onError(final Throwable t) {
      */
     static final class OffloadedCancellable implements Cancellable {
         private final Cancellable cancellable;
+        private final BooleanSupplier shouldOffload;
         private final Executor executor;
 
-        OffloadedCancellable(final Cancellable cancellable, final Executor executor) {
+        OffloadedCancellable(final Cancellable cancellable,
+                             final BooleanSupplier shouldOffload, final Executor executor) {
             this.cancellable = requireNonNull(cancellable);
+            this.shouldOffload = shouldOffload;
             this.executor = executor;
         }
 
         @Override
         public void cancel() {
+            boolean offloadCancel;
             try {
-                LOGGER.debug("executing Completable cancel on {}", executor);
-                executor.execute(() -> safeCancel(cancellable));
+                offloadCancel = shouldOffload.getAsBoolean();
             } catch (Throwable t) {
-                LOGGER.warn("Failed to execute task on the executor {}. " +
-                                "Invoking Cancellable (cancel()) in the caller thread. Cancellable {}. ",
-                        executor, cancellable, t);
-                // As a policy, we call the target in the calling thread when the executor is inadequately
-                // provisioned. In the future we could make this configurable.
+                // As a policy, we call the target in the calling thread when the "hint" fails. In the future we could
+                // make this configurable.
+                offloadCancel = false;
```

Comment:
lets play it safe here and offload. if folks are blocking and we skip offloading this may lead to bad times (blocking IO thread, increase latency, etc.), if folks are non-blocking and we offload they will have to use the thread pool (assumed less impact).

Reply:
OK.